### PR TITLE
Genome browser files ADR

### PR DESCRIPTION
**Pull request by ens-ds23**
Genome browser files ADR

This is a new ADR for the genome browser, detailing the decision to use files (at least as an interim).

## Type

- [ ] Bug fix
- [X] New feature
- [ ] Improvement to existing feature

## Related JIRA Issue(s)
ENSWBSITES-383

## Importance
No due date. Not urgent.

## Description
The genome browser uses file at the moment for its backing data. Along with the already exisiting ADR 0006, this document describes why.

Essentially, the design of the genome-browser calls for a file-like API (at least a ranomly-accessible byte-stream) and while many technologies provide that, actual files are perhaps the simplest and have good performance out-of-the-box.

## Views affected
Genome browser

### Other effects
None

- [ ] I have checked any requirements for Google Analytics
- [ ] I have created any required tests
- [ ] The PR has as its final commit a WASM/JS update commit if any Rust files were updated.

## Possible complications
None envisaged

### CODE REVIEW

File: src/assets/browser/app/doc/manual/docs/arch/0007-files.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+## Consequence of
+
+None
+
+## Context
+
+One of the ways that the genome browser is intended to be fast is that the design includes extensive preprocessing at the server end. Much of the slowness from the current site comes from the immense amount of work required in some cases to determine the colour of even a single pixel. Such pixels are always the same colour and to merely record the colour saves an awful lot of time.
+
+On the other hand, our images are slightly more dynamid that map tiles, so some intermeidate form is required which records the pos-processed information. As illustrated by ADR 0006, there are additional benefits to recording information in a heavily denormalised format, as close to visual as practically possible.
```
Comment (azangru) [inline src/assets/browser/app/doc/manual/docs/arch/0007-files.md:17]: typo: dynamid -> dynamic

File: src/assets/browser/app/doc/manual/docs/arch/0007-files.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+The delivery for this data should:
+
+1. support the access pattern of looking up data along a one-dimensional genomic axis, along with a number of discrete axes;
+2. be CDN friendly;
+3. be easily cacheable;
+4. support data-blobs (a consequence of using a bytecode, see ADR0006).
+
+There are a number of solutions in this area. As well as desirables for the final product, we should take care not to expend more energy than necessary in this area, particularly while we are experimenting.
+
+There are a number of file common bioinformatics formats which handle this kind of access pattern reasonably cleanly. Major advantages include:
```
Comment (azangru) [inline src/assets/browser/app/doc/manual/docs/arch/0007-files.md:30]: typo `file common bioinformatics formats` -> `common bioinformatics file formats` ?

File: src/assets/browser/app/doc/manual/docs/arch/0007-files.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+## Decision
+
+The cheapness of the implementation makes files easy to start with even if we entertain the possibility of changing later. Filesystems are not a good way of organising the different discrete axes, so a database to manage these visible through a service would make a lot of sense. This would greatly reduce the need to rely on filenames and directory structure. This can be developed independently.
+
+As long as a technology can emulate a "big" file in providing random-access to a byte-array with indexing with efficient caching, there is no reason not to migrate to it in time, particularly if it helps solve the issue of the discrete axes. There are a lot of systems in this space. The API would remain the familiar file-like API.
+
+To allow this to happen, we must be careful not to rely on ideosyncracies of the POSIX file API (permissions, locks, owners, etc), but restrict ourselves to a minimal easily-migratable subset. Essentially, anything that supports an HTTP GET with Byte-Range requests efficiently (and does so with effective caching) should be a plausible candidate.
```
Comment (azangru) [inline src/assets/browser/app/doc/manual/docs/arch/0007-files.md:49]: typo ideosyncracies  -> idiosyncrasies

File: src/assets/browser/app/doc/manual/docs/arch/0007-files.md (hunk 1 above)
Comment (ens-ma7) [reply src/assets/browser/app/doc/manual/docs/arch/0007-files.md:17]: Another typo intermeidate -> intermediate

File: src/assets/browser/app/doc/manual/docs/arch/0007-files.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+1. They are typically heavily modelled and not easily adapted to pre-processed blobs;
+2. They don't handle the discrete axes (multiple datasets) well and we are left instead with a mess on the filesystem;
+3. they don't fit well with the rest of the system. (However, we've seen that this is fundamentally different data to the rest of the system and with different access patterns, so that's not overly surprising).
+
+## Decision
+
+The cheapness of the implementation makes files easy to start with even if we entertain the possibility of changing later. Filesystems are not a good way of organising the different discrete axes, so a database to manage these visible through a service would make a lot of sense. This would greatly reduce the need to rely on filenames and directory structure. This can be developed independently.
+
+As long as a technology can emulate a "big" file in providing random-access to a byte-array with indexing with efficient caching, there is no reason not to migrate to it in time, particularly if it helps solve the issue of the discrete axes. There are a lot of systems in this space. The API would remain the familiar file-like API.
```
Comment (ens-ma7) [inline src/assets/browser/app/doc/manual/docs/arch/0007-files.md:47]: with indexing with efficient caching -> with indexing and efficient caching

**Comment by ens-ds23** [comment]
The latest commit should have the typos and spelling issues above fixed.

**Comment by ens-ds23** [comment]
Thanks I missed that. I've updated that now. I'll leave it local and if there's anything more coming in will roll it into that.

File: src/assets/browser/app/doc/manual/docs/arch/0007-files.md [hunk 5]
```diff
@@ -27,7 +27,7 @@ The delivery for this data should:
 
 There are a number of solutions in this area. As well as desirables for the final product, we should take care not to expend more energy than necessary in this area, particularly while we are experimenting.
 
-There are a number of file common bioinformatics formats which handle this kind of access pattern reasonably cleanly. Major advantages include:
+There are a number of file common bioinformatics file formats which handle this kind of access pattern reasonably cleanly. Major advantages include:
```
Comment (msmanoj) [inline src/assets/browser/app/doc/manual/docs/arch/0007-files.md:30]: Should this be "number of common bioinformatics file formats" ?

**Comment by ens-ma7** [comment]
@ens-ds23 is it ok to merge this? It has been approved and doesn't look like there is any more changes.

**Pull request merged**